File: leetcode_hot_100/leetcode_100.java
import java.util.*;

//package leetcode_hot_100;
public class leetcode_100 {
    // hash
    public static int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> ht = new HashMap<>();
        for (int i = 0; i < nums.length; i++) {
            if (ht.containsKey(target - nums[i])) {
                return new int[]{ht.get(target - nums[i]), i};
            }
            ht.put(nums[i], i);
        }
        return new int[0];
    }
    // two pointer
    public static int maxArea(int[] height) {
        int ans = 0, left = 0, right = height.length - 1;
        while (left < right) {
            int area = (right - left) * Math.min(height[left], height[right]);
            ans = Math.max(ans, area);
            if (height[left] < height[right]) ++left;
            else --right;
        }
        return ans;
    }
    // sliding window
    public static int lengthOfLongestSubstring(String S) {
        char[] s = S.toCharArray();
        int n = s.length, ans = 0, left = 0;
        boolean[] has = new boolean[128];
        for (int right = 0; right < n; right++) {
            char c = s[right];
            while (has[c]) {
                has[s[left++]] = false;
            }
            has[c] = true;
            ans = Math.max(ans, right - left + 1);
        }
        return ans;
    }

    // subarray
    public static int subarraySum(int[] nums, int k) {
        int ans = 0, s = 0;
        Map<Integer, Integer> cnt = new HashMap<>(nums.length + 1);
        cnt.put(0, 1);
        for (int x : nums) {
            s += x;
            ans += cnt.getOrDefault(s - k, 0);
            cnt.merge(s, 1, Integer::sum);
        }
        return ans;
    }

    // array
    public static int maxSubArray(int[] nums) {
        int ans = Integer.MIN_VALUE;
        int f = 0;
        for (int x: nums) {
            f = Math.max(f, 0) + x;
            ans = Math.max(ans, f);
        }
        return ans;
    }

    // matrix
    public static int[][] setZeroes(int[][] matrix) {
        int m = matrix.length, n = matrix[0].length;
        boolean[] row = new boolean[m];
        boolean[] col = new boolean[n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (matrix[i][j] == 0) {
                    row[i] = col[j] = true;
                }
            }
        }
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (row[i] || col[j]) {
                    matrix[i][j] = 0;
                }
            }
        }
        return matrix;
    }

    // graph
    static void dfs(String[][] grid, int r, int c) {
        int nr = grid.length;
        int nc = grid[0].length;
        if (r < 0 || c < 0 || r >= nr || c >= nc || grid[r][c] == "0") {
            return;
        }
        grid[r][c] = "0";
        dfs(grid, r - 1, c);
        dfs(grid, r + 1, c);
        dfs(grid, r, c - 1);
        dfs(grid, r, c + 1);
    }
    public static int numsIslands(String[][] grid) {
        if (grid == null || grid.length == 0) {
            return 0;
        }
        int nr = grid.length;
        int nc = grid[0].length;
        int num_islands = 0;
        for (int r = 0; r < nr; ++r) {
            for (int c = 0; c < nc; ++c) {
                if (grid[r][c] == "1") {
                    ++num_islands;
                    dfs(grid, r, c);
                }
            }
        }
        return num_islands;
    }

    // backtrace
    private static int[] nums;
    private static List<Integer> path;
    private static boolean[] onPath;
    private static final List<List<Integer>> ans = new ArrayList<>();

    public List<List<Integer>> permute(int[] nums) {
        this.nums = nums;
        path = Arrays.asList(new Integer[nums.length]);
        onPath = new boolean[nums.length];
        dfs(0);
        return ans;
    }

    private static void dfs(int i) {
        if (i == nums.length) {
            ans.add(new ArrayList<>(path));
            return;
        }
        for (int j = 0; j < nums.length; ++j) {
            if (!onPath[j]) {
                path.set(i, nums[j]);
                onPath[j] = true;
                dfs(i + 1);
                onPath[j] = false; // 恢复现场
            }
        }
    }

    // bisect
    public int searchInsert(int[] nums, int target) {
//        return Arrays.binarySearch(nums, target);
        int left = 0, right = nums.length - 1;
        while (left <= right) {
            int mid = left + (right - left); // 2;
            if (nums[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return left;
    }

    // stack
    public boolean isValid(String s) {
        if (s.length() % 2 != 0) { // s 长度必须是偶数
            return false;
        }
        Map<Character, Character> mp = new HashMap<Character, Character>() {{
            put(')', '(');
            put(']', '[');
            put('}', '{');
        }};
        Deque<Character> st = new ArrayDeque<>();
        for (char c : s.toCharArray()) {
            if (!mp.containsKey(c)) { // c 是左括号
                st.push(c); // 入栈
            } else if (st.isEmpty() || st.pop() != mp.get(c)) { // c 是右括号
                return false; // 没有左括号，或者左括号类型不对
            }
        }
        return st.isEmpty(); // 所有左括号必须匹配完毕
    }

    // queue
    public int findKthLargest(int[] nums, int k) {
        int heapSize = nums.length;
        buildMaxHeap(nums, heapSize);
        for (int i = nums.length - 1; i >= nums.length - k + 1; --i) {
            swap(nums, 0, i);
            --heapSize;
            maxHeapify(nums, 0, heapSize);
        }
        return nums[0];
    }
    public void buildMaxHeap(int[] a, int heapSize) {
        for (int i = heapSize / 2; i >= 0; --i) {
            maxHeapify(a, i, heapSize);
        }
    }
    public void maxHeapify(int[] a, int i, int heapSize) {
        int l = i * 2 + 1, r = i * 2 + 2, largest = i;
        if (l < heapSize && a[l] > a[largest]) {
            largest = l;
        }
        if (r < heapSize && a[r] > a[largest]) {
            largest = r;
        }
        if (largest != i) {
            swap(a, i, largest);
            maxHeapify(a, largest, heapSize);
        }
    }
    public void swap(int[] a, int i, int j) {
        int temp = a[i];
        a[i] = a[j];
        a[j] = temp;
    }
    // greedy

    // dp

    // bit

    // listnode
    class ListNode {
        int val;
        ListNode next;
        ListNode() {};
        ListNode(int val) {this.val = val;}
        ListNode(int val, ListNode next) {this.val = val; this.next = next; }
    }
    public ListNode arr2ListNode(int[] nums) {
        ListNode head = new ListNode(nums[0]);
        ListNode tmp = head;
        for (int i = 1; i < nums.length; ++i) {
            tmp.next = new ListNode(nums[i]);
            tmp = tmp.next;
        }
        return head;
    }
    public ListNode reverseList(ListNode head) {
        ListNode pre = null, cur = head;
        while (cur != null) {
            ListNode nxt = cur.next;
            cur.next = pre;
            pre = cur;
            cur = nxt;
        }
        return pre;
    }
    // treenode
    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {};
        TreeNode(int val) {this.val = val;}
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    TreeNode arr2TreeNodeHelper(int[] nums, int start, int end) {
        if (start > end) return null;
        int mid = (start + end) >> 1;
        TreeNode root = new TreeNode(nums[mid]);
        root.left = arr2TreeNodeHelper(nums, start, mid - 1);
        root.right = arr2TreeNodeHelper(nums, mid + 1, end);
        return root;
    }
    TreeNode arr2TreeNode(int[] nums) {
        return arr2TreeNodeHelper(nums, 0, nums.length - 1);
    }

    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        inorder(root, res);
        return res;
    }
    public void inorder(TreeNode root, List<Integer> res) {
        if (root == null) {
            return;
        }
        inorder(root.left, res);
        res.add(root.val);
        inorder(root.right, res);
    }
    public static void main(String[] args) {
        int[] nums1 = {2,7,11,15};
        System.out.println(Arrays.toString(twoSum(nums1, 9)));
        int[] height11 = {1,8,6,2,5,4,8,3,7};
        System.out.println(maxArea(height11));
        String s3 = "abcabcbb";
        System.out.println(lengthOfLongestSubstring(s3));
        int[] nums560 = {1,1,1};
        System.out.println(subarraySum(nums560, 2));
        int[] nums53 = {-2,1,-3,4,-1,2,1,-5,4};
        System.out.println(maxSubArray(nums53));
        int[][] matrix73 = {{1,1,1},{1,0,1},{1,1,1}};
        Arrays.stream(setZeroes(matrix73)).map(item -> Arrays.toString(item)).forEach(System.out::println);
        String[][] grid200 = {{"1","1","1","1","0"},{"1","1","0","1","0"},{"1","1","0","0","0"},{"0","0","0","0","0"}};
        System.out.println(numsIslands(grid200));
        leetcode_100 l = new leetcode_100();
        int[] nums46 = {1,2,3};
        System.out.println(l.permute(nums46));
        int[] nums35 = {1,3,5,6};
        System.out.println(l.searchInsert(nums35, 5));
        String s20 = "{{}}()";
        System.out.println(l.isValid(s20));
        int[] nums215 = {3,2,1,5,6,4};
        System.out.println(l.findKthLargest(nums215, 2));

        int[] nums206 = {1,2,3,4,5};
        ListNode t = l.arr2ListNode(nums206);
        System.out.println(l.reverseList(t).val);

        int[] nums94 = {1,0, 2, 3};
        TreeNode t1 = l.arr2TreeNode(nums94);
        System.out.println(l.inorderTraversal(t1));


//        System.out.println("hi");
    }

}

File: leetcode_hot_100/leetcode_100.go
package main

import (
	"fmt"
	"math"
)

// hash
func twoSum(nums []int, target int) []int {
	ht := map[int]int{}
	for i, x := range nums {
		if p, ok := ht[target-x]; ok {
			return []int{p, i}
		}
		ht[x] = i
	}
	return nil
}

// two pointer
func maxArea(height []int) (ans int) {
	left, right := 0, len(height)-1
	for left < right {
		area := (right - left) * min(height[left], height[right])
		ans = max(ans, area)
		if height[left] < height[right] {
			left++
		} else {
			right--
		}
	}
	return
}

// slide window
func lengthOfLongestSubstring(s string) (ans int) {
	window := [128]bool{}
	left := 0
	for right, c := range s {
		for window[c] {
			window[s[left]] = false
			left++
		}
		window[c] = true
		ans = max(ans, right-left+1)
	}
	return
}

// subarray
func subarraySum(nums []int, k int) (ans int) {
	s := 0
	cnt := map[int]int{0: 1}
	for _, x := range nums {
		s += x
		ans += cnt[s-k]
		cnt[s]++
	}
	return
}

// array
func maxSubArray(nums []int) int {
	ans := math.MinInt
	minPreSum := 0
	preSum := 0
	for _, x := range nums {
		preSum += x
		ans = max(ans, preSum-minPreSum)
		minPreSum = min(minPreSum, preSum)
	}
	return ans
}

// matrix

// graph

// backtrace
func permute(nums []int) (ans [][]int) {
	n := len(nums)
	path := make([]int, n)
	onPath := make([]bool, n)
	var dfs func(int)
	dfs = func(i int) {
		if i == n {
			ans = append(ans, append([]int(nil), path...))
			return
		}
		for j, on := range onPath {
			if !on {
				path[i] = nums[j]
				onPath[j] = true
				dfs(i + 1)
				onPath[j] = false
			}
		}
	}
	dfs(0)
	return
}

// bisect
func searchInsert(nums []int, target int) int {
	left, right := 0, len(nums)-1
	for left <= right {
		mid := left + (right-left)>>1
		if nums[mid] < target {
			left = mid + 1
		} else {
			right = mid - 1
		}
	}
	return left
}

// stack
func isValid(s string) bool {
	if len(s)%2 != 0 {
		return false
	}
	mp := map[rune]rune{')': '(', ']': '[', '}': '{'}
	st := []rune{}
	for _, c := range s {
		if mp[c] == 0 {
			st = append(st, c)
		} else {
			if len(st) == 0 || st[len(st)-1] != mp[c] {
				return false
			}
			st = st[:len(st)-1]
		}
	}
	return len(st) == 0
}

// queue
func findKthLargest(nums []int, k int) int {
	heapify(nums)
	for i := 0; i < k-1; i++ {
		pop(&nums)
	}
	return nums[0]
}
func heapify(nums []int) {
	n := len(nums)
	for i := n/2 - 1; i >= 0; i-- {
		down(nums, i, n)
	}
}
func down(nums []int, i, n int) bool {
	parent := i
	for {
		left := 2*parent + 1
		if left >= n || left < 1 {
			break
		}
		max := left
		if right := left + 1; right < n && nums[right] > nums[left] {
			max = right
		}
		if nums[max] < nums[parent] {
			break
		}
		nums[parent], nums[max] = nums[max], nums[parent]
		parent = max
	}
	return parent > i
}
func pop(nums *[]int) int {
	last := len(*nums) - 1
	(*nums)[0], (*nums)[last] = (*nums)[last], (*nums)[0]
	down(*nums, 0, last)
	rst := (*nums)[last]
	(*nums) = (*nums)[:last]
	return rst
}

// greedy
func maxProfit(prices []int) (ans int) {
	minPrice := prices[0]
	for _, p := range prices {
		ans = max(ans, p-minPrice)
		minPrice = min(minPrice, p)
	}
	return
}

// dp
func climbStars(n int) int {
	memo := make([]int, n+1)
	var dfs func(int) int
	dfs = func(i int) int {
		if i <= 1 { // 递归边界
			return 1
		}
		p := &memo[i]
		if *p != 0 { // 之前计算过
			return *p
		}
		res := dfs(i-1) + dfs(i-2)
		*p = res // 记忆化
		return res
	}
	return dfs(n)
}

func climbStairs_(n int) int {
	f := make([]int, n+1)
	f[0], f[1] = 1, 1
	for i := 2; i <= n; i++ {
		f[i] = f[i-1] + f[i-2]
	}
	return f[n]
}

// bit

// listnode
type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseList(head *ListNode) *ListNode {
	var pre, cur *ListNode = nil, head
	for cur != nil {
		nxt := cur.Next
		cur.Next = pre
		pre = cur
		cur = nxt
	}
	return pre
}

// treenode
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func inorderTraversal(root *TreeNode) (res []int) {
	var inorder func(node *TreeNode)
	inorder = func(node *TreeNode) {
		if node == nil {
			return
		}
		inorder(node.Left)
		res = append(res, node.Val)
		inorder(node.Right)
	}
	inorder(root)
	return
}

func main() {
	nums1 := []int{2, 7, 11, 15}
	fmt.Println(twoSum(nums1, 9))
	height11 := []int{1, 8, 6, 2, 5, 4, 8, 3, 7}
	fmt.Println(maxArea(height11))
	s3 := "abcabcbb"
	fmt.Println(lengthOfLongestSubstring(s3))
	nums560 := []int{1, 2, 3}
	fmt.Println(subarraySum(nums560, 3))
	nums53 := []int{-2, 1, -3, 4, -1, 2, 1, -5, 4}
	fmt.Println(maxSubArray(nums53))
	nums46 := []int{1, 2, 3}
	fmt.Println(permute(nums46))
	nums35 := []int{1, 3, 5, 6}
	fmt.Println(searchInsert(nums35, 5))
	s20 := "()[]{}"
	fmt.Println(isValid(s20))
	nums215 := []int{3, 2, 1, 5, 6, 4}
	fmt.Println(findKthLargest(nums215, 2))
	nums121 := []int{7, 1, 5, 3, 6, 4}
	fmt.Println(maxProfit(nums121))
	fmt.Println(climbStairs_(3))
	fmt.Println(climbStars(3))

	nums206 := []int{1, 2, 3, 4, 5}
	var arr2listNode = func(nums []int) *ListNode {
		if len(nums) == 0 {
			return nil
		}
		head := &ListNode{Val: nums[0]}
		tmp := head
		for i := 1; i < len(nums); i++ {
			tmp.Next = &ListNode{Val: nums[i]}
			tmp = tmp.Next
		}
		return head
	}
	l := arr2listNode(nums206)
	fmt.Println(reverseList(l).Val)

	nums94 := []int{1, 2, 3}
	arr2treeNode := func(nums []int) *TreeNode {
		if len(nums) == 0 {
			return nil
		}
		var arr2treeNodeHelper func(nums []int, start int, end int) *TreeNode
		arr2treeNodeHelper = func(nums []int, start int, end int) *TreeNode {
			if start > end {
				return nil
			}
			mid := (start + end) / 2
			node := &TreeNode{Val: nums[mid]}
			node.Left = arr2treeNodeHelper(nums, start, mid-1)
			node.Right = arr2treeNodeHelper(nums, mid+1, end)
			return node
		}
		return arr2treeNodeHelper(nums, 0, len(nums)-1)
	}
	r := arr2treeNode(nums94)
	fmt.Println(inorderTraversal(r))
}
